server/src/main/java/dev/slimevr/tracking/processor/skeleton/LegTweaks.java

```diff
@@ -713,6 +757,178 @@ private void correctUnlockedRightFootTracker() {
+			float weightL = 0.0f;
+			float weightR = 0.0f;
+
+			// the further from the ground the foot is, the less weight it
+			// should have
+			weightL = getFootPlantWeight(leftFootPos);
+			weightR = getFootPlantWeight(rightFootPos);
+
+			// perform the correction
+			leftFootRotation
+				.set(
+					leftFootRotation
+						.slerp(
+							leftFootRotation,
+							isolateYaw(leftFootRotation),
+							weightL * masterWeightL
+						)
+				);
+
+			rightFootRotation
+				.set(
+					rightFootRotation
+						.slerp(
+							rightFootRotation,
+							isolateYaw(rightFootRotation),
+							weightR * masterWeightR
+						)
+				);
+		}
+
+		// corrects rotations when the foot is in the air by rotating the foot
+		// down so that the toes are touching
+		if (toeSnap) {
+			// this correction step has its own weight vars
+			float weightL = 0.0f;
+			float weightR = 0.0f;
+
+			// first compute the angle of the foot
+			float angleL = getToeSnapAngle(leftFootPos);
+			float angleR = getToeSnapAngle(rightFootPos);
+
+			// then compute the weight of the correction
+			weightL = getToeSnapWeight(leftFootPos, angleL);
+			weightR = getToeSnapWeight(rightFootPos, angleR);
+
+			// depending on the state variables, the correction weights should
+			// be clamped
+			if (!leftToeTouched) {
+				weightL = Math.min(weightL, leftToeAngle);
+			}
+			if (!rightToeTouched) {
+				weightR = Math.min(weightR, rightToeAngle);
+			}
+
+			// then slerp the rotation to the new rotation based on the weight
+			leftFootRotation
+				.slerp(
+					leftFootRotation,
+					replacePitch(leftFootRotation, -angleL),
+					weightL * masterWeightL
+				);
+			rightFootRotation
+				.slerp(
+					rightFootRotation,
+					replacePitch(rightFootRotation, -angleR),
+					weightR * masterWeightR
+				);
+
+			// update state variables regarding toe snap
+			if (leftFootPos.y - floorLevel > footLength * MAXIMUM_TOE_DOWN_ANGLE) {
+				leftToeTouched = false;
+				leftToeAngle = weightL;
+			} else if (leftFootPos.y - floorLevel < 0.0f) {
+				leftToeTouched = true;
+				leftToeAngle = 1.0f;
+			}
+			if (rightFootPos.y - floorLevel > footLength * MAXIMUM_TOE_DOWN_ANGLE) {
+				rightToeTouched = false;
+				rightToeAngle = weightR;
+			} else if (rightFootPos.y - floorLevel < 0.0f) {
+				rightToeTouched = true;
+				rightToeAngle = 1.0f;
+			}
+		}
+
+		// update the foot rotations in the buffer
+		bufferHead.setLeftFootRotationCorrected(leftFootRotation);
+		bufferHead.setRightFootRotationCorrected(rightFootRotation);
+
+		// finally update the skeletons rotations with the new rotations
+		skeleton.computedLeftFootTracker.rotation.set(leftFootRotation);
+		skeleton.computedRightFootTracker.rotation.set(rightFootRotation);
+	}
+
+	// returns the length of the xz components of the normalized difference
+	// between two vectors
+	public float getXZAmount(Vector3f vec1, Vector3f vec2) {
+		return vec1
+			.subtract(vec2)
+			.normalize()
```

Comment:
`normalizeLocal()` can be used here instead, as `subtract()` is creating a new `Vector3f` object. I also see potential for using a buffer `Vector3f` as a method argument for this calculation, especially since it isn't ending up returning a `Vector3f` in the end.
```suggestion
			.normalizeLocal()
```

Reply:
I'm not sure that I know what using a buffer  `Vector3f` would look like, I'd be happy to do that though if you could explain it

Reply:
We can chat on Discord later or i can make a commit for you if you want

Reply:
I'm happy to chat about it!